nat-lab/tests/utils/ping.py

```diff
@@ -52,7 +52,6 @@ async def execute(self) -> None:
         await self._process.execute(stdout_callback=self.on_stdout)
 
     async def wait_for_next_ping(self, timeout: Optional[float] = None) -> None:
-        self._next_ping_event.clear()
```

Comment:
I think this will break this situation:
```python
async with Ping(<connection>, <ping_target_ip_address>).run() as ping:
           await ping.wait_for_next_ping()
           await do_some_other_stuff_for_some_time()
           await break_network()
           await fix_network();
           await ping.wait_for_next_ping()
```

The second `wait_for_next_ping` will succeed regardless if the `fix_network()` actually worked - there will be messages in the buffer from the time `do_some_other_stuff_for_some_time` was running.

Reply:
yeah, I specifically checked all uses of `Ping(..)` to verify that this is not occurring. Although it may easy to introduce such scenario.

Maybe to make the interface "cleaner" we should merge `.run()` and `.wait_for_next_ping()`? :thinking:

Reply:
I think something should be added to this PR that will make it impossible to make such a mistake. I like the idea of merging those 2 methods but I worry that it might quite some time to do. If that is the case, maybe a cheaper way to go would be to stop/kill the ping process the first time the wait_for_next_ping is called and/or throw an exception the second time it is called.

Reply:
what if you were to make `run()` wait for first ping like this (not touching the code of `ping.wait_for_next_ping()`:
```
    @asynccontextmanager
    async def run(self) -> AsyncIterator["Ping"]:
        async with self._process.run(stdout_callback=self.on_stdout):
            await self._next_ping_event.wait()
            yield self
```

and then in the code we would do something like:
```
async with Ping(<connection>, <ping_target_ip_address>).run() as ping:
           await ping
```
and in the other scenarios, where @tomaszklak mentioned, we could do:
```
async with Ping(<connection>, <ping_target_ip_address>).run() as ping:
           await ping
           await do_some_other_stuff_for_some_time()
           await break_network()
           await fix_network();
           await ping.wait_for_next_ping()
```
**Note** Haven't tested it myself, not sure atm if `await ping` wouldn't just break async manager, but just an idea :thinking:

Reply:
I did some refactoring :point_right: :point_left: